mina.netty/src/main/java/org/kaazing/mina/core/service/AbstractIoAcceptorEx.java

```diff
@@ -59,20 +59,21 @@ public BindFuture bindAsync(final SocketAddress localAddress) {
         if (getHandler() == null) {
             throw new IllegalStateException("handler is not set.");
         }
+
         BindFuture bound = bindAsyncInternal(localAddress);
         bound.addListener(new IoFutureListener<BindFuture>() {
             @Override
             public void operationComplete(BindFuture future) {
                 if (future.isBound()) {
                     boolean activate = false;
-                    synchronized (bindLock) {
+                    synchronized (boundAddresses) {
                         if (boundAddresses.isEmpty()) {
                             activate = true;
                         }
                         boundAddresses.add(localAddress);
-                        if (activate) {
-                            getListeners().fireServiceActivated();
-                        }
+                    }
+                    if (activate) {
```

Comment:
The problem about moving the firing of the listeners outside the synchronized block is the outcome (service activated or not activated) may now be wrong if there's a race between bind and unbind done by different threads: if this the first bind and bind sync block completes before unbind sync block but unbind thread first serviceDisactivated before bind thread fires serviceActivated then outcome will be service activated, which is wrong.